### DISCUSSION TITLE
Arduino 2

### DISCUSSION THREAD

**Discussion by lerouxb** [SpenceKonde/megaTinyCore#355] [Future Development Discussion]
So.. have you thought about adding support for the new Arduino IDE v2? I'm wondering if debug over UPDI would be possible to add?

**Comment by SpenceKonde** [comment]
Nope! Have given zero thought to Arduino 2. I have a very VERY long list of action items that people are waiting for.

I will do what is necessary to make the core compatible with new versions of the IDE if someone (likely @per1234 ) tells me what I need to do. I dearly hope that making basic functionality work does not require a great deal of conversion effort!

I, for one, have never wanted or needed hardware debugging

**Reply by lerouxb** [reply]
I'm a "log stuff and never fire up a debugger" kind of programmer myself. Was just thinking that since I always have the programmer plugged in anyway it could do away with the need to always have serial connected. But I suspect I'd always have serial connected because it is so convenient anyway.

**Reply by per1234** [reply]
I specifically targeted the DrAzzy and MCUdude platforms in my Arduino IDE 2.x testing work because I know they are essential parts of the Arduino ecosystem. I only have one megaTinyCore-compatible board (ATtiny817 Xplained Mini) so I didn't get the best coverage of uploads, but I think it should be fine. Please submit a bug report to [the arduino/arduino-ide repo](https://github.com/arduino/arduino-ide/issues) or let me know if you find any problems.

---
As for feature parity with the classic Arduino IDE, megaTinyCore works just fine as is with one exception. The ["Upload Using Programmer" by default](https://arduino.github.io/arduino-cli/latest/platform-specification/#upload-using-programmer-by-default) capability is currently not supported (https://github.com/arduino/arduino-ide/issues/103). Inconvenient, but you can just do a **Sketch > Upload Using Programmer** instead so it's not a show stopper. 

---
As for the extended functionality, unfortunately the picture is not so pretty:

### Language server

The [clangd language server](https://clangd.llvm.org/) that's the engine of [the Arduino Language Server](https://github.com/arduino/arduino-language-server) seems to be lacking support for the megaAVR 0-series and tinyAVR 0-series and 1-series (AKA "megaAVR") (https://github.com/arduino/arduino-ide/issues/31). This means that the autocomplete, editor hover, "Go To Definition", "Go To Symbol", "Peek Definition" features don't work for megaTinyCore, MegaCoreX, Arduino megaAVR Boards, and DxCore as well. The language server is cool, but not essential. This also affects the official Uno WiFi Rev2 and Nano Every boards, so this bug hits home for Arduino as well. If anyone knows anything about this situation with clangd, I'd be very interested to hear about it. I did some searching without any luck, but I don't know much about this so I might have been using the wrong keywords.

## Debugger

The VS Code extension used for the integrated debugger is advertised as providing:
https://github.com/Marus/cortex-debug#readme
>Debugging support for ARM Cortex-M Microcontrollers

So that doesn't sound very promising.  But there was something interesting here:
https://github.com/Marus/cortex-debug/issues/13#issuecomment-381001391
>As for supporting other MCUs that are not Cortex-M based - I'm not really opposed to the idea.

Unfortunately, it doesn't look like there have been any developments in that area since then.

There is a forum thread specifically about this as it relates to the ATtiny1614, so you might want to follow that in case any interesting new information arises:
https://forum.arduino.cc/index.php?topic=730288

There is also some interesting information about an investigation into setting up debugging for the Arduino Uno here:
https://github.com/arduino/arduino-ide/issues/87

I actually haven't made any use of the debugger yet for a project, but I've been having a ton of fun playing around with it on my SAMD boards. I would really love to see more platforms taking advantage of this new capability. I had a shot at using AVaRICE and avr-gdb a while back and failed completely, so if debugging for AVR could somehow receive the "Arduino treatment", that would be super awesome.

**Reply by SpenceKonde** [reply]
Wow! Lovely, thanks for the information there. Best of all, it sounds like I don't need to do anything. That's the best kind of news (humor aside, I already have enough tasks to keep me busy for years)

The one thing I'd be most worried about is whether the Serial-Adapter-and-resistor upload (which uses a different tool for the upload, a python script that makes use of Microchip's pymcuprog utility). If you'd like, I'd be happy to send you a  serial adapter with the added resistor that is known to work this way, a 3p programming cable, and a target board or two, I'd be happy to, just drop me an email :-)

**Reply by per1234** [reply]
I've used the "pyupdi", jtag2updi, and [microUPDI](https://github.com/MCUdude/microUPDI) programmers to upload to the ATtiny817 on both Windows and Linux without having any problems.

Ironically, I haven't been able to manage an upload via the onboard mEDBG provided by Microchip (["avrdude: jtag3_edbg_recv(): Inconsistent fragment number; expect 1, got 0"](https://github.com/SpenceKonde/megaTinyCore/issues/312#issuecomment-777386697)), but I have the same problem with the classic Arduino IDE, so it's nothing specific to Arduino IDE 2.x and my only use for the board is testing so I'm not concerned with getting that working.

---
One other issue I should mention, since it likely affects megaTinyCore development and beta tester workflows, is that Arduino IDE 2.x doesn't currently support boards platforms manually installed in the `hardware` subfolder of the sketchbook. Currently, only Boards Manager installations are recognized (https://github.com/arduino/arduino-ide/issues/126). That will require some enhancements to Arduino CLI (which does support the sketchbook platforms, but doesn't provide the information in the necessary form via the gRPC interface to Arduino IDE 2.x), and that is in progress now. I think this will be supported in an upcoming beta release.
